### PR TITLE
[resharding] Refactor `build_state_for_split_shards` and related functions in one file

### PR DESCRIPTION
**Pull request by shreyan-gupta**
PR to refactor all resharding related code into one file `resharding.rs`

All future resharding related code should hopefully live here.

### CODE REVIEW

File: chain/chain/src/chain.rs [hunk 1]
```diff
@@ -425,7 +424,7 @@ type BlockApplyChunksResult = (CryptoHash, Vec<Result<ApplyChunkResult, Error>>)
 /// Facade to the blockchain block processing and storage.
 /// Provides current view on the state according to the chain state.
 pub struct Chain {
-    store: ChainStore,
+    pub(crate) store: ChainStore,
```
Comment (wacban) [inline chain/chain/src/chain.rs:427]: There is a store() method on Chain, can you use that instead?

File: chain/chain/src/resharding.rs [hunk 2]
```diff
@@ -0,0 +1,182 @@
+/// Implementation for all resharding logic.
+/// StateSplitRequest and StateSplitResponse are exchanged across the client_actor and sync_jobs_actor.
+/// build_state_for_split_shards_preprocessing and build_state_for_split_shards_postprocessing are handled
+/// by the client_actor while the heavy resharding build_state_for_split_shards is done by sync_jobs_actor
+/// so as to not affect client.
```
Comment (wacban) [inline chain/chain/src/resharding.rs:5]: <3

File: chain/chain/src/resharding.rs [hunk 3]
```diff
@@ -0,0 +1,182 @@
+/// Implementation for all resharding logic.
+/// StateSplitRequest and StateSplitResponse are exchanged across the client_actor and sync_jobs_actor.
+/// build_state_for_split_shards_preprocessing and build_state_for_split_shards_postprocessing are handled
+/// by the client_actor while the heavy resharding build_state_for_split_shards is done by sync_jobs_actor
```
Comment (wacban) [inline chain/chain/src/resharding.rs:4]: nit: SyncJobsActor

File: chain/chain/src/resharding.rs [hunk 4]
```diff
@@ -0,0 +1,182 @@
...
+use crate::types::RuntimeAdapter;
+use crate::Chain;
+
+#[derive(actix::Message)]
+#[rtype(result = "()")]
+pub struct StateSplitRequest {
+    pub runtime_adapter: Arc<dyn RuntimeAdapter>,
+    pub sync_hash: CryptoHash,
+    pub shard_id: ShardId,
+    pub shard_uid: ShardUId,
```
Comment (wacban) [inline chain/chain/src/resharding.rs:32]: seems a bit redundant (shard_uid contains shard_id) it's something like before and after resharding
anyway let's not change anything in this PR and keep it as pure refactor but if you have a moment you can see what's up with this

File: chain/chain/src/resharding.rs [hunk 5]
```diff
@@ -0,0 +1,182 @@
...
+    orig_state_root: StateRoot,
+    state_roots: HashMap<ShardUId, StateRoot>,
+    account_id_to_shard_id: &(dyn Fn(&AccountId) -> ShardUId + 'a),
+) -> Result<HashMap<ShardUId, StateRoot>, Error> {
+    let orig_trie_update = tries.new_trie_update_view(orig_shard_uid, orig_state_root);
+
+    let mut start_index = None;
+    let mut new_state_roots = state_roots;
+    while let Some((next_index, receipts)) =
+        get_delayed_receipts(&orig_trie_update, start_index, STATE_PART_MEMORY_LIMIT)?
```
Comment (wacban) [inline chain/chain/src/resharding.rs:58]: cool syntax though hard to parse at first sight

File: chain/chain/src/resharding.rs [hunk 6]
```diff
@@ -0,0 +1,182 @@
...
+        )?;
+        new_state_roots = updated_state_roots;
+        start_index = Some(next_index);
+        store_update.commit()?;
+    }
+
+    Ok(new_state_roots)
+}
+
+impl Chain {
```
Comment (wacban) [inline chain/chain/src/resharding.rs:73]: cool I like the idea of splitting chain impl across multiple files

File: chain/chain/src/resharding.rs [hunk 7]
```diff
@@ -0,0 +1,182 @@
...
+    pub fn build_state_for_split_shards(
+        state_split_request: StateSplitRequest,
+    ) -> Result<HashMap<ShardUId, StateRoot>, Error> {
+        let StateSplitRequest {
+            runtime_adapter,
+            shard_uid,
+            state_root,
+            next_epoch_shard_layout,
+            state_split_status,
+            ..
```
Comment (wacban) [inline chain/chain/src/resharding.rs:114]: strange that we don't unpack all args 
ditto - not to be fixed here but would be nice to understand what's up

File: chain/client/src/sync_jobs_actor.rs [hunk 8]
```diff
@@ -135,20 +137,13 @@ impl actix::Handler<WithSpanContext<StateSplitRequest>> for SyncJobsActor {
...
         let (_span, msg) = handler_debug_span!(target: "client", msg);
-        let results = msg.runtime_adapter.build_state_for_split_shards(
-            msg.shard_uid,
-            &msg.state_root,
-            &msg.next_epoch_shard_layout,
-            msg.state_split_status,
-        );
+        let shard_id = msg.shard_id;
+        let sync_hash = msg.sync_hash;
+        let results = Chain::build_state_for_split_shards(msg);
```
Comment (wacban) [inline chain/client/src/sync_jobs_actor.rs:142]: nit: new_state_roots then you can omit the the name in line 145

**Review by wacban — APPROVED** [review]
Brilliant, thanks!

File: chain/chain/src/resharding.rs (hunk 4 above)
Comment (shreyan-gupta) [reply chain/chain/src/resharding.rs:32]: Yeah, I honestly don't know why is it done this way. Probably an oversight in the original implementation.

File: chain/chain/src/resharding.rs (hunk 7 above)
Comment (shreyan-gupta) [reply chain/chain/src/resharding.rs:114]: The arguments are used in functions that call build_state_for_split_shards as part of response I think.

File: chain/chain/src/chain.rs (hunk 1 above)
Comment (shreyan-gupta) [reply chain/chain/src/chain.rs:427]: Aah, there was a `mut_store()` method that I had missed. That's why it wasn't working before.

File: chain/chain/src/resharding.rs (hunk 5 above)
Comment (shreyan-gupta) [reply chain/chain/src/resharding.rs:58]: Haha! I'd done the same in `core/store/src/trie/from_flat.rs`
